### PR TITLE
[#766] Added modern entity-field parser with 'field_parser' configuration flag.

### LINKED ISSUE #642: Escaping in createNode() changed with PHP 8.1
I'm using RawNodeContext::nodeCreate() to create a bundle that has an entity reference. This worked in Drupal 8/PHP 7. When upgrading to Drupal 9/PHP 8.1 my behats break.

As title of the entity reference to create in the bundle I used "0000000000 \- Enterprise". The actual title of the reference is "0000000000 - Behat enterprise". When I use the title without the "\" I get a "number of bound variables does not match number of tokens" error which I also had in Drupal 8/PHP 7. The difference is that "0000000000 \- Enterprise" now results in a query:

```
SELECT "base_table"."vid" AS "vid", "base_table"."nid" AS "nid"
FROM
"node" "base_table"
INNER JOIN "node_field_data" "node_field_data" ON "node_field_data"."nid" = "base_table"."nid"
│ WHERE ("node_field_data"."title" LIKE '0000000000 \\\\- Behat enterprise' ESCAPE '\\') AND ("node_field_data"."type" IN ('enterprise'))
```
4 \'s in the query for the title of the reference instead of the 2 \'s I would expect. The query with the 4 \'s doesn't match anything, when I manually change the query to 2 \'s I get the data I expect. 

Because of PHP 8 something seems to double escape.

### LINKED ISSUE #766: Replace inline compound field syntax to resolve separator conflicts
## Summary

The current inline compound syntax in `parseEntityFields()` uses ` - ` (space-dash-space) as a column separator and `: ` (colon-space) as a key-value delimiter. Both characters appear frequently in natural-language field values, causing parsing failures that have no workaround.

This issue proposes replacing the inline compound syntax in the next major version with an unambiguous alternative.

### 1. The ` - ` separator breaks entity references with dashes in titles

When an entity reference field value contains ` - `, `parseEntityFields()` splits it into compound columns instead of treating it as a single label.

**Example:** A page titled `Alpha - Bravo` cannot be referenced:

```gherkin
Given "page" content:
  | title         |
  | Alpha - Bravo |
When I am viewing a "post" content:
  | title                | My post       |
  | field_post_reference | Alpha - Bravo |
```

`parseEntityFields()` splits `Alpha - Bravo` into `['Alpha', 'Bravo']`. The `EntityReferenceHandler` receives an array where it expects a string, causing `Drupal\Core\Database\InvalidQueryException`.

There is **no escape mechanism** for ` - `. Users cannot quote it, backslash-escape it, or work around it in any way.

See: https://github.com/jhedstrom/drupalextension/issues/642

### 2. The `: ` key-value delimiter breaks values containing colons

Any field value containing `: ` (colon-space) triggers named-column parsing inside a compound group.

**Example:** A text field with value `Note: this is important` inside a compound split would be parsed as key `Note` with value `this is important`.

### 3. Three independent separator characters create combinatorial conflicts

The current DSL uses three separators that must all be avoided in field values:

| Separator | Purpose | Conflicts with |
|-----------|---------|---------------|
| `,` | Multi-value | Commas in titles, tags, addresses |
| ` - ` | Compound columns | Dashes in titles, descriptions, URLs |
| `: ` | Named columns | Colons in labels, time values, descriptions |

Only `,` has an escape mechanism (CSV double-quoting). The other two have none.

### 4. The `,` escape mechanism interacts poorly with ` - `

Even though commas can be escaped with double quotes (`"Tag, three"`), double-quoting does **not** prevent ` - ` splitting. A value like `"Alpha - Bravo"` has its quotes stripped by `str_getcsv()` before the ` - ` split runs, so the split still happens.

## Current syntax reference

```gherkin
# Multi-value (comma-separated)
| field_tags           | Tag one, Tag two, "Tag, three"              |

# Compound positional columns (` - ` separated)
| field_post_links     | Link 1 - http://example.com                 |

# Compound named columns (` - ` and `: ` separated)
| field_post_address   | country: BE - locality: Brussel - postal_code: 1000 |

# Multi-value compound
| field_post_links     | Link 1 - http://example.com, Link 2 - http://example.com |

# Multicolumn headers (already unambiguous)
| field_post_address:country      | BE       |
| :locality                       | Brussel  |
| :postal_code                    | 1000     |
```

## Proposed alternatives

Below are four candidate DSL designs. All preserve comma-separated multi-value and the existing `field:column` header syntax. They differ only in how compound columns are expressed **inline** (within a single cell).

### Option A — Bracket groups

Use `[brackets]` for compound columns. Named columns use `key:value` (no space) inside brackets.

```gherkin
| field_post_reference | Alpha - Bravo                                               |
| field_post_links     | [Link 1, http://example.com], [Link 2, http://example.com]  |
| field_post_address   | [country:BE, locality:Brussel, thoroughfare:Louisalaan 1, postal_code:1000] |
| field_tags           | Tag one, Tag two, "Tag, three"                              |
```

- Simple values stay clean — no separator conflicts
- Brackets clearly delineate compound structures
- Commas, dashes, colons all work in plain values
- Risk: brackets in literal values would need escaping (rare)

### Option B — Inline JSON for compounds

Use JSON objects `{...}` for compound columns. Plain values are unchanged. The parser detects JSON by checking if a trimmed value starts with `{`.

```gherkin
| field_post_reference | Alpha - Bravo                                |
| field_post_links     | {"title":"Link 1","uri":"http://example.com"}, {"title":"Link 2","uri":"http://example.com"} |
| field_post_address   | {"country":"BE","locality":"Brussel","thoroughfare":"Louisalaan 1","postal_code":"1000"} |
| field_tags           | Tag one, Tag two, "Tag, three"               |
```

- Zero ambiguity — JSON is self-describing
- Developers already know JSON
- `json_decode()` handles all escaping and edge cases
- Verbose for compound fields

### Option C — Typed bracket tokens

Bracket-prefixed tokens signal a processing mode. Extensible for future field types.

```gherkin
| field_post_reference | Alpha - Bravo                                |
| field_post_links     | [uri:http://example.com title:Link 1], [uri:http://example.com title:Link 2] |
| field_post_address   | [country:BE locality:Brussel thoroughfare:Louisalaan 1 postal_code:1000] |
| field_post_date      | [relative:-1 week]                           |
| field_tags           | Tag one, Tag two, "Tag, three"               |
```

- Extensible — new token types without changing the parser core
- Plain values are completely clean
- Space-separated `key:value` inside brackets may need quoting for values with spaces

### Option D — Full JSON

All structured values use JSON arrays and objects.

```gherkin
| field_post_reference | ["Alpha - Bravo"]                            |
| field_post_links     | [{"title":"Link 1","uri":"http://example.com"}, {"title":"Link 2","uri":"http://example.com"}] |
| field_post_address   | {"country":"BE","locality":"Brussel","thoroughfare":"Louisalaan 1","postal_code":"1000"} |
| field_tags           | ["Tag one", "Tag two", "Tag, three"]         |
```

- Completely unambiguous for all cases
- Too verbose for simple multi-value fields like tags
- Not very "Gherkin-like"

## Migration considerations

- The `field:column` header syntax is unaffected and should be preserved as-is.
- A deprecation period should support both old and new syntax with warnings.
- A migration script or Rector rule could automate conversion of existing feature files.
- The multi-value comma separator and `"quoted"` escaping should be preserved — they work well.

## Related issues

- https://github.com/jhedstrom/drupalextension/issues/642 — original report of entity reference escaping failure

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
Closes #766

## Summary

Introduces a new `EntityFieldParser` that replaces the legacy field-value parsing logic in `RawDrupalContext::parseEntityFields()`. The new parser uses a single uniform escape mechanism (double quotes), detects compound mode by value form rather than separator spacing, and surfaces clear `ParseException` / `MultipleParseException` errors instead of returning silently malformed data. A new `field_parser` configuration flag in `behat.yml` lets projects opt into the legacy parser during migration; the legacy parser emits a deprecation notice once per process and is removed entirely in 6.1.

This also resolves #642 as a consequence: entity reference titles that contain ` - ` (e.g. `"My article - Part 2"`) no longer require any escaping in scalar mode, because dashes are treated as literal text and are not interpreted as compound separators by the modern parser.

### New parser (`src/Drupal/DrupalExtension/Parser/`)

- `EntityFieldParser` - full rewrite of the field-value grammar with named column support, quoted-scalar detection, token syntax (`[relative:-1 week]`), and unambiguous compound detection via `key:"value"` pairs.
- `Exception/ParseException` - structured single-value error with file/line context for use in step-definition error messages.
- `Exception/MultipleParseException` - aggregates parse errors across a table so all failures surface in one shot.

### Context changes (`src/Drupal/DrupalExtension/Context/RawDrupalContext.php`)

- `parseEntityFields()` delegates to whichever parser is selected by `field_parser` config.
- New `getFieldParser()` factory reads the config flag and returns either `EntityFieldParser` or `LegacyEntityFieldParser` (with a one-time deprecation notice when `legacy` is chosen).

### Configuration (`src/Drupal/DrupalExtension/ServiceContainer/DrupalExtension.php`)

- `field_parser` key added to the extension's config tree (values: `default` | `legacy`).

### Test coverage

- `tests/phpunit/src/Parser/EntityFieldParserTest.php` - 441 assertions covering all grammar branches, error paths, and edge cases.
- `tests/behat/features/field_handlers.feature` - updated to use modern syntax.
- `tests/behat/features/field_handlers_legacy.feature` - new file; exercises the legacy parser under the `drupal_legacy_parser` Behat profile.

### Migration docs

- `MIGRATION.md` - new "Field syntax" section with side-by-side syntax table and named-column reference.

### Parser selection flow

```
behat.yml: field_parser: default | legacy
                |
                v
     RawDrupalContext::parseEntityFields()
                |
                v
     RawDrupalContext::getFieldParser($entity_type, $classifier)
                |
                +-- field_parser == 'legacy'  ---> LegacyEntityFieldParser
                |   (also emits deprecation notice once per process)
                |
                +-- otherwise                 ---> EntityFieldParser
```

### Syntax comparison (selected cases)

```
                          Legacy                 Modern
                          ------                 ------
Compound single:          country: BE            country:"BE", locality:"Brussel"
                          - locality: Brussel

Scalar with dash:         "Alpha - Bravo"        Alpha - Bravo   (no escape needed)
   (resolves #642)        (workaround required)

Scalar with semicolon:    Hello; world           "Hello; world"

Token at compound pos:    (not expressible)      value:[relative:-1 week]
```

## Deprecation timeline

- **6.0 (this PR)**: `field_parser: legacy` is opt-in; selecting it emits a deprecation notice once per process.
- **6.1**: `field_parser: legacy` is removed; setting it produces a hard configuration error.

## Related

- https://github.com/jhedstrom/DrupalDriver/issues/365 - tracks future work in `drupal/drupal-driver` to formalize creation-hint property aliases; no driver change is required for this PR, but the issue is informational for field-type discoverability.

## Summary by CodeRabbit

* **New Features**
  * Enhanced field-value parsing with structured key/value syntax, improved escape handling, and character-offset error reporting that aggregates multiple issues per cell.

* **Documentation**
  * Added migration guidance and parser reference documenting modern syntax, escape rules, whitespace semantics, and error-reporting behavior.

* **Configuration**
  * New parser selection option (modern default) with a deprecated legacy mode slated for removal.

* **Tests**
  * Expanded and reorganized test/feature suites to cover modern and legacy parsing, edge cases, and aggregated error reporting.

### CODE REVIEW

File: MIGRATION.md [hunk 1]
```diff
@@ -35,6 +35,105 @@ Update each occurrence in your `.feature` files.
...
+The modern parser reports parse errors with character-level position
+information. A typical failure looks like:
+
+```
+Parse error in field_post_address:
+country:"BE", locality:Brussel, postal_code:"1000"
+              ^
+unquoted_compound_value at offset 14: Compound column "locality" must use a quoted string or token.
+Hint: Wrap the value in double quotes or use a [token:value] form.
+```
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply MIGRATION.md]: Fixed in 245b2983. Added 'text' language identifier to both fenced code blocks.

File: src/Drupal/DrupalExtension/Parser/EntityFieldParser.php [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+   * {@inheritdoc}
+   */
+  public function parse(array $values): array {
+    $multicolumn_field = '';
+    $multicolumn_column = '';
+    $multicolumn_fields = [];
+    $parsed = [];
+    $errors = [];
+    $error_cell = '';
+
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Drupal/DrupalExtension/Parser/EntityFieldParser.php]: Fixed in 245b2983. The aggregate exception now uses errors[0]->cell so the cell text matches the error code/offset that the parent constructor inherits from the first error.

File: src/Drupal/DrupalExtension/Context/RawDrupalContext.php [hunk 3]
```diff
@@ -395,11 +396,27 @@ public function parseEntityFields(EntityStubInterface $stub, array $ignored_prop
...
-    return new LegacyEntityFieldParser($entity_type, $classifier);
+    $mode = $this->getDrupalParameter('field_parser') ?? 'default';
+
+    if ($mode === 'legacy') {
+      static $deprecation_emitted = FALSE;
+
+      if (!$deprecation_emitted) {
+        // phpcs:ignore Drupal.Semantics.FunctionTriggerError
+        @trigger_error('The legacy field parser is deprecated and will be removed in 6.1. Remove "field_parser: legacy" from your behat.yml to migrate. See MIGRATION.md.', E_USER_DEPRECATED);
+        $deprecation_emitted = TRUE;
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Drupal/DrupalExtension/Context/RawDrupalContext.php:417]: Not applicable here. The Drupal coding standard enforced by this project (Drupal.Semantics.UnsilencedDeprecation) explicitly requires the @ prefix on trigger_error() for E_USER_DEPRECATED. Removing it produces a lint error: 'All trigger_error calls used for deprecation must be prefixed by an @'. The convention is intentional - default PHP error handling suppresses the notice as noise during normal operation, while frameworks that subscribe to deprecations (Symfony deprecation collector, Drupal's own handler) still receive it via set_error_handler. The notice is intended to be visible to those subscribers, not to default PHP output.

[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Drupal/DrupalExtension/Context/RawDrupalContext.php:417]: Correcting my earlier response - this is a valid concern and is fixed in 1566694d. Behat does not run with a Symfony-style deprecation collector that observes suppressed notices, so the `@` would hide the message from test authors. The unsilenced `trigger_error` is correct here. The Drupal lint rule is locally suppressed via `phpcs:ignore`.

File: src/Drupal/DrupalExtension/Parser/EntityFieldParser.php [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+      return [$key, $value];
+    }
+
+    throw new ParseException(
+      'unquoted_compound_value',
+      0,
+      $cell,
+      sprintf('Compound column "%s" must use a quoted string or token.', $key),
+      'Wrap the value in double quotes or use a [token:value] form.',
+    );
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Drupal/DrupalExtension/Parser/EntityFieldParser.php:461]: Fixed in 5581f995. `splitTopLevel` now returns segment+offset tuples. `parseRecord` and `parseColumn` thread an absolute `base_offset` through. `readQuotedString` and `readToken` accept the original cell and absolute base offset so all thrown errors point to the actual position in the source cell rather than the fragment.
